### ISSUE TITLE
Better performace through dart:FFI

### ISSUE THREAD

**Issue by dvagala** [zezo357/pytorch_lite#14]
Hey, I've made an example [here](https://github.com/dvagala/pytorch-flutter-FFI-example) of how to add C++ LibTorch to Flutter and call it directly through dart:FFI. My measures suggest that the inference is significantly faster than using the method channels, and moreover, all the C++ code can be shared with Android and iOS, so no duplication anymore.

I'm just posting it here if anyone wants to play with it or make a PR.

**Comment by zezo357** [comment]
I tried to compile pytorch to use ffi but failed, I would love to implement it in this package

**Comment by zezo357** [comment]
Can you make a pr with your changes? This will benefit the package greatly.

Thank you.

**Comment by dvagala** [comment]
@zezo357 yeah I can relate, I've also spent few days trying to compile pytorch from source to get the `.a` static files, but always hit a dead end. Eventually, I had a success with using dynamic `.so` that comes bundled in the [original pytorch package](https://mvnrepository.com/artifact/org.pytorch/pytorch_android/1.13.0). For the iOS, there's cocoapod for LibTorch, so this part was easy.

I think I don't get the answer about performance. You mean if using the `ffigen` will increase the performance? Honestly, I'm not familiar with the `ffigen`, but just a quick look at pubdev and it says is's just for generating the binding for the C++ code, and I wrote that layer manually. So see no reason for any performance difference. 

Sorry I currently don't have time to make a PR for this package. Though in that [example](https://github.com/dvagala/pytorch-flutter-FFI-example) I linked here, I've made a precise tutorial on how I've made it work, and it's not that difficult. Feel free to ask anything in case of a problem

**Comment by zezo357** [comment]
> @zezo357 yeah I can relate, I've also spent few days trying to compile pytorch from source to get the `.a` static files, but always hit a dead end. Eventually, I had a success with using dynamic `.so` that comes bundled in the [original pytorch package](https://mvnrepository.com/artifact/org.pytorch/pytorch_android/1.13.0). For the iOS, there's cocoapod for LibTorch, so this part was easy.
> 
> I think I don't get the answer about performance. You mean if using the `ffigen` will increase the performance? Honestly, I'm not familiar with the `ffigen`, but just a quick look at pubdev and it says is's just for generating the binding for the C++ code, and I wrote that layer manually. So see no reason for any performance difference. 
> 
> Sorry I currently don't have time to make a PR for this package. Though in that [example](https://github.com/dvagala/pytorch-flutter-FFI-example) I linked here, I've made a precise tutorial on how I've made it work, and it's not that difficult. Feel free to ask anything in case of a problem

Can I ask where were you able to get the Aten folder? If those files are found somewhere? 

Because if those files can be found a can compile pytorch to make it work cross platform using ffi

**Comment by dvagala** [comment]
> Can I ask where were you able to get the Aten folder? If those files are found somewhere?

As I said, for Android you can get all required files (including Aten headers) in the [original pytorch package](https://mvnrepository.com/artifact/org.pytorch/pytorch_android/1.13.0) specifically in the [aar](https://repo1.maven.org/maven2/org/pytorch/pytorch_android/1.13.0/pytorch_android-1.13.0.aar) file. You just unzip this `.aar` and there take content of `jni/` and `headers/`. But as I mentioned in that [example](https://github.com/dvagala/pytorch-flutter-FFI-example) you can find all of this.

**Comment by zezo357** [comment]
> > Can I ask where were you able to get the Aten folder? If those files are found somewhere?
> 
> As I said, for Android you can get all required files (including Aten headers) in the [original pytorch package](https://mvnrepository.com/artifact/org.pytorch/pytorch_android/1.13.0) specifically in the [aar](https://repo1.maven.org/maven2/org/pytorch/pytorch_android/1.13.0/pytorch_android-1.13.0.aar) file. You just unzip this `.aar` and there take content of `jni/` and `headers/`. But as I mentioned in that [example](https://github.com/dvagala/pytorch-flutter-FFI-example) you can find all of this.

thank you very much, i will try to do it

**Comment by zezo357** [comment]
@dvagala i am sry but i couldnt find the native_pytorch/CMakeLists.txt file to run :(

can you check why its not there?

**Comment by zezo357** [comment]
## 3.0.0
* Converted to work on ffi (for improved performance, and ios support)
